colossalai/nn/optimizer/distributed_adafactor.py

```diff
@@ -0,0 +1,439 @@
+            param_shape: Paramater Shape of Each params;
+            use_zero: Whether or not to use zero;
+
+        """
+        self.tensor_parallel_group = tensor_parallel_group  # "Expected row process group"
+        self.data_parallel_group = data_parallel_group
+        if self.tensor_parallel_group is not None:
+            self.tensor_parallel_size = dist.get_world_size(self.tensor_parallel_group)
+        if self.data_parallel_group is not None:
+            self.data_parallel_size = dist.get_world_size(self.data_parallel_group)
+        self.use_zero = use_zero
+
+        self.shard_to_param = shard_to_param if shard_to_param is not None else {}
+        # grad is None, cause we dont setup now
+        for group in self.param_groups:
+            for p in group["params"]:
+                self.param_is_dtensor_dict[id(p)] = is_distributed_tensor(self.shard_to_param.get(id(p)))
+                self.grad_shape_dict[id(p)] = self.shard_to_param.get(id(p)).shape
+                self.factored_dict[id(p)], self.use_first_moment_dict[id(p)] = self._get_options(
+                    group, self.grad_shape_dict[id(p)]
+                )
+                if self.param_is_dtensor_dict[id(p)]:
+                    self.shard_spec_dict[id(p)] = get_sharding_spec(self.shard_to_param.get(id(p)))
+                else:
+                    self.shard_spec_dict[id(p)] = None
+
+    @staticmethod
+    def _get_lr(param_group, param_state):
+        rel_step_sz = param_group["lr"]
+        if param_group["relative_step"]:
+            min_step = 1e-6 * param_state["step"] if param_group["warmup_init"] else 1e-2
+            rel_step_sz = min(min_step, 1.0 / math.sqrt(param_state["step"]))
+        param_scale = 1.0
+        if param_group["scale_parameter"]:
+            param_scale = max(param_group["eps"][1], param_state["RMS"])
+        return param_scale * rel_step_sz
+
+    @staticmethod
+    def _get_options(param_group, param_shape):
+        """
+        Determines whether the current param is factored
+        Args:
+            param_group : param group
+            param_shape : Original Shape of param
+
+        """
+        factored = len(param_shape) >= 2
+        use_first_moment = param_group["beta1"] is not None
+        return factored, use_first_moment
+
+    @staticmethod
+    def _rms(tensor, param_is_dtensor, use_zero, tp_size, dp_size, tp_group, dp_group):
+        tensor_sum = tensor.pow(2).sum()
+        num_of_element = tensor.numel()
+
+        if param_is_dtensor:
+            # reduce tensor_sum  from tp_group
+            dist.all_reduce(tensor_sum, group=tp_group)
+            num_of_element = num_of_element * tp_size
+            if use_zero:
+                dist.all_reduce(tensor_sum, group=dp_group)
+                num_of_element = num_of_element * dp_size
+        else:
+            if use_zero:
+                dist.all_reduce(tensor_sum, group=dp_group)
+                num_of_element = num_of_element * dp_size
```

Comment:
This can be simplified as `if use_zero: xxxx`

Reply:
> This can be simplified as `if use_zero: xxxx`


Fixed in https://github.com/hpcaitech/ColossalAI/pull/5484/commits/0a7f68269077347655d5c2bc284320bac29d1116;